### ISSUE TITLE
Unable to correctly print PDF file: Error: Ignoring spurious ET operator.

### ISSUE THREAD

**Issue by paulmenzel** [OpenPrinting/cups-filters#93]
I am using Debian Sid/unstable and cups-filters 1.21.6-3.

The university document [antrag_exmatrikulation.pdf](https://github.com/OpenPrinting/cups-filters/files/2881078/antrag_exmatrikulation.pdf) causes problems on different printers.

Printing on the Canon iR-ADV C2225i using the Canon PCL, an error page is first printed, and on the first page after *Email* everything is missing. The second page is fine.

     **** Error: Ignoring spurious ET operator. Output may be inco ****

For that printer a special filter `sicgsfilter` has to be used, and is distributed by Canon. This also happens with their CQue 4.0-1 release from September 2018.

     *FoomaticRIPCommandLine: "gs -q -dBATCH %F -dPARANOIDSAFER -dNOPAUSE %B%A%C -dPDFFitPage  -sOutputFile=- - %D%E | sicgsfilter -MPCL -NP %G%H%I -u&user; -V&quot;&title;&quot; -n&copies; "

On the Epson printer *WorkForce WF-2660*, no error page is printed, but on the first page everything after *Email* *is missing* too.

Trying to use different pdftops renderers did not make any difference. For example,

     lpr -o pdftops-renderer=pdftops -P chipprinterpcl antrag_exmatrikulation.pdf

gave the same result, the same with `pdftocairo`, `acroread`, `mupdf`.

Can you reproduce this problem with the printers you have access to?

Suggestions on how to further debug this are much appreciated.

**Comment by paulmenzel** [comment]
Some more things, I forgot to mention. Doing `pdf2ps` manually, and printing that, works.

The Epson model is *WorkForce WF-2660*.

Following the Debian Wiki article [DissectingandDebuggingtheCUPSPrintingSystem](https://wiki.debian.org/DissectingandDebuggingtheCUPSPrintingSystem), and creating a printer actually printing to a file, the file is created, but I am unable to view it.

Lastly, using that method to analyze the differences of the pdftops renderer, there is none.

```
$ lpr -o pdftops-renderer=mupdf -P chipprinterfilepcl ~/antrag_exmatrikulation.pdf 
$ sudo chown pmenzel: /tmp/test-pcl.ps
$ cp -a /tmp/test-pcl.ps test-pcl-mupdf.ps
$ lpr -o pdftops-renderer=pdftops -P chipprinterfilepcl ~/antrag_exmatrikulation.pdf 
$ sudo chown pmenzel: /tmp/test-pcl.ps
$ cp -a /tmp/test-pcl.ps test-pcl-pdftops.ps
$ lpr -o pdftops-renderer=gs -P chipprinterfilepcl ~/antrag_exmatrikulation.pdf 
$ sudo chown pmenzel: /tmp/test*
$ cp -a /tmp/test-pcl.ps test-pcl-gs.ps
$ ls -ltr test*.ps
insgesamt 6732
-rw------- 1 pmenzel pmenzel 1292628 Feb 19 20:43 test-pcl-pdftops.ps
-rw------- 1 pmenzel pmenzel 1292628 Feb 19 20:43 test-pcl-mupdf.ps
-rw------- 1 pmenzel pmenzel 1292628 Feb 19 20:43 test-pcl-gs.ps
$ md5sum test*.ps
175a9585a0beeab02a08dc032880d40d  test-pcl-gs.ps
175a9585a0beeab02a08dc032880d40d  test-pcl-mupdf.ps
175a9585a0beeab02a08dc032880d40d  test-pcl-pdftops.ps
```

Here is the PPD file (with suffix `.txt` for GitHub).

*   [chipprinterpcl.ppd.txt](https://github.com/OpenPrinting/cups-filters/files/2881759/chipprinterpcl.ppd.txt)

**Comment by paulmenzel** [comment]
Actually [looking at the Postscript output][1] – I thought I had already –, the errors are in there.

```
ESC%-12345X@PJL CQue4.0-1 x86_64
@PJL COMMENT CANPJL SET USERNAME="pmenzel@localhost"
@PJL COMMENT CANPJL SET DOCNAME="antrag_exmatrikulation.pdf!pmenz"
@PJL SET RENDERMODE=COLOR
@PJL COMMENT CANPJL SET FORCEMONOCHROME=FALSE
@PJL COMMENT CANPJL SET SORTERMODE=GENERICON
@PJL COMMENT CANPJL SET STAPLE=GENERICOFF
@PJL COMMENT CANPJL SET BOOKLET=GENERICOFF
@PJL SET DENSITY=0
@PJL ENTER LANGUAGE=PCL
   **** Error: Ignoring spurious ET operator.
               Output may be incorrect.
   **** Error: Executing Do inside a text block, attempting to recover
               Output may be incorrect.
   **** Error: Executing Do inside a text block, attempting to recover
               Output may be incorrect.
   **** Error: Ignoring spurious ET operator.
               Output may be incorrect.
   **** Error: Executing Do inside a text block, attempting to recover
               Output may be incorrect.
   **** Error: illegal nested BT operator detected.
               Output may be incorrect.
   **** Error: Ignoring spurious ET operator.
               Output may be incorrect.
   **** Error: Executing Do inside a text block, attempting to recover
               Output may be incorrect.
   **** Error: illegal nested BT operator detected.
               Output may be incorrect.
ESCEESC&l1S…
[…]
````

[1]: https://wiki.ubuntu.com/DebuggingPrintingProblems

**Comment by paulmenzel** [comment]
Editing the PPD, and leaving just the Ghostscript command in there, it can still be reproduced.

```
-*FoomaticRIPCommandLine: "gs -q -dBATCH %F -dPARANOIDSAFER -dNOPAUSE %B%A%C -dPDFFitPage  -sOutputFile=- - %D%E | sicgsfilter -MPCL -NP %G%H%I -u&user; -V&quot;&title;&quot; -n&copies; "
+*FoomaticRIPCommandLine: "gs -q -dBATCH %F -dPARANOIDSAFER -dNOPAUSE %B%A%C -dPDFFitPage  -sOutputFile=- - %D%E "
```

This goes in line, that this is also happening on the Epson printer.

```
$ head -18 /tmp/test-pcl.ps
   **** Error: Ignoring spurious ET operator.
               Output may be incorrect.
   **** Error: Executing Do inside a text block, attempting to recover
               Output may be incorrect.
   **** Error: Executing Do inside a text block, attempting to recover
               Output may be incorrect.
   **** Error: Ignoring spurious ET operator.
               Output may be incorrect.
   **** Error: Executing Do inside a text block, attempting to recover
               Output may be incorrect.
   **** Error: illegal nested BT operator detected.
               Output may be incorrect.
   **** Error: Ignoring spurious ET operator.
               Output may be incorrect.
   **** Error: Executing Do inside a text block, attempting to recover
               Output may be incorrect.
   **** Error: illegal nested BT operator detected.
               Output may be incorrect.
```

**Comment by tillkamppeter** [comment]
Also reported [here](https://bugs.launchpad.net/bugs/1817033).

**Comment by tillkamppeter** [comment]
The PDF file contains a fillable form. Did you fill it in and save the PDF with filled form before printing? Or did you print the original PDF without filling the form?

**Comment by tillkamppeter** [comment]
Can you update your Debian unstable so that you get cups-filters 1.22.x? Does this solve your problem?

**Comment by paulmenzel** [comment]
On 02/21/19 12:10, Till Kamppeter wrote:
> The PDF file contains a fillable form. Did you fill it in and save
> the PDF with filled form before printing? Or did you print the
> original PDF without filling the form?
I printed it without filling the form (from Evince and `lpr`).

**Comment by paulmenzel** [comment]
On 02/21/19 12:11, Till Kamppeter wrote:
> Can you update your Debian unstable so that you get cups-filters
> 1.22.x? Does this solve your problem?

Adding the experimental archive under `/etc/apt/sources*`, and
running `apt install cups-filters/experimental`, it prints fine
now. Sorry for not trying the latest release in the first place.

**Comment by paulmenzel** [comment]
[So, the commits mentioned in merge/pull request #88 to use QPDF seem to work around the issue.](
https://github.com/OpenPrinting/cups-filters/pull/88#issuecomment-454167373)

The Ghostscript developers seem to be interested in fixing this, too, but I haven’t figured out, how to reproduce this with plain Ghostscript commands. Could you help me here?

**Comment by paulmenzel** [comment]
Just a quick update, that updating to Cairo 1.17.2 also fixes the problem, as it includes [commit 4c8813f0 (pdf: add missing flush)](https://gitlab.freedesktop.org/cairo/cairo/commit/4c8813f0eaacc32c27126ad2296951a626300b89) fixing issue [Cairo generates incorrect PDF files](https://gitlab.freedesktop.org/cairo/cairo/issues/342). Thank you to Tobias Hoffmann for mentioning that in the Launchpad bug report.

Distributions should probably back port that commit. I’ll ask the Cairo developers to cherry-pick this into the 1.16 branch.

**Comment by paulmenzel** [comment]
For the record, here the commands to directly reproduce the issue with the example PDF file.

With Cairo 1.15.12 and Cairo 1.16.0:

```
dev/shm$ pdftocairo -pdf ~/antrag_exmatrikulation.pdf antrag_exmatrikulation.pdf
dev/shm$ gs -o repaired.pdf -sDEVICE=pdfwrite -dPDFSETTINGS=/prepress antrag_exmatrikulation.pdf 
GPL Ghostscript 9.26 (2018-11-20)
Copyright (C) 2018 Artifex Software, Inc.  All rights reserved.
This software comes with NO WARRANTY: see the file PUBLIC for details.
Processing pages 1 through 2.
Page 1
   **** Error: Ignoring spurious ET operator.
               Output may be incorrect.
   **** Error: Executing Do inside a text block, attempting to recover
               Output may be incorrect.
   **** Error: Executing Do inside a text block, attempting to recover
               Output may be incorrect.
   **** Error: Ignoring spurious ET operator.
               Output may be incorrect.
   **** Error: Executing Do inside a text block, attempting to recover
               Output may be incorrect.
   **** Error: illegal nested BT operator detected.
               Output may be incorrect.
   **** Error: Ignoring spurious ET operator.
               Output may be incorrect.
   **** Error: Executing Do inside a text block, attempting to recover
               Output may be incorrect.
   **** Error: illegal nested BT operator detected.
               Output may be incorrect.
Page 2
```
